### PR TITLE
Kokonoe

### PR DESCRIPTION
**Pull request by daniel-k-taylor**

### CODE REVIEW

File: scenes/game/local_game.gd [hunk 1]
```diff
@@ -5560,6 +5837,154 @@ func strike_send_attack_to_discard_or_gauge(performing_player : Player, card):
...
+	# The player is mid-strike and playing an extra attack card.
+	# These are the steps of extra attack:
+	# * Pay its cost.
+	# * Put it into play, gaining any Armor, Guard, and passive effects on it.
+	# * Perform its Before effects, then check its Range.
+	# * If it hit, perform its Hit effects.
+	# * If it hit, deal damage equal to its printed Power.
+	# * Perform its After effects.
+	# * At the end of the Strike, if the additional attack hit, add it to your Gauge. Otherwise, discard it.
+	# * Note that bonuses and effects that apply to your attack (e.g., Continuous Boost effects such as "+2 Power") only apply to your initial attack unless they specify "Your attacks have [...]" (as the Graviton does).
```
Comment (Mach131) [inline scenes/game/local_game.gd:5851]: hopefully this isn't something we have to find and support for other boosts retroactively?

File: scenes/game/local_game.gd [hunk 2]
```diff
@@ -5560,6 +5837,154 @@ func strike_send_attack_to_discard_or_gauge(performing_player : Player, card):
...
+				if game_over:
+					change_game_state(Enums.GameState.GameState_GameOver)
+				else:
+					# Intentional events = because events are passed in.
+					events = continue_resolve_strike(events)
+				break
+	return events
+
+func determine_if_attack_hits(events, attacker_player : Player, defender_player : Player, card : GameCard):
+	# events actually is used, but there's an unused parameter warning for some reason, so add _.
```
Comment (Mach131) [inline scenes/game/local_game.gd:5962]: is this an outdated comment?

File: test/unit/test_randomai.gd [hunk 3]
```diff
@@ -1,7 +1,7 @@
 extends GutTest
 
 # Leave at 1 checked in so someone doesn't accidentally run all tests at 100.
-const RandomIterations = 1
+const RandomIterations = 0
```
Comment (Mach131) [inline test/unit/test_randomai.gd:4]: reminder to set to 1

File: test/unit/test_randomai.gd [hunk 4]
```diff
@@ -1,7 +1,7 @@
 extends GutTest
```
Comment (Mach131) [inline test/unit/test_randomai.gd:1]: This currently just sets up a bunch of mirror matches, right? I wonder if it would be helpful to add a special test that runs two different random characters against each other; it might help to find a few more strange corner cases, if annotated properly

**Review by Mach131 — APPROVED** [review]
Seems reasonable to me if the tests pass, just a bunch of text/formatting-related comments mostly; very hard to envision the exact ways in which things might fall apart, hence the randomai tests suggestion

File: data/card_definitions.json [hunk 5]
```diff
@@ -6914,6 +6914,416 @@
...
+		"speed": 4,
+		"power": 4,
+		"armor": 0,
+		"guard": 0,
+		"effects": [
+			{
+				"timing": "before",
+				"effect_type": "advance",
+				"stop_on_buddy_space": "gravitron",
+				"description": "Advance 4. If you enter Gravitron's space, stop and gain advantage.",
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply data/card_definitions.json:7299]: no, so i added that

File: data/decks/kokonoe.json [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+	"on_exceed": {
+		"effect_type": "take_bonus_actions",
+		"amount": 1,
+		"and": {
+			"effect_type": "place_buddy_in_any_space",
+			"buddy_name": "Gravitron"
+		}
+	},
+	"overdrive_effect": {
+		"effect_type": "place_buddy_in_any_space",
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply data/decks/kokonoe.json:19]: the overdrive one isn't, so i'll make it a choice

File: scenes/game/local_game.gd [hunk 7]
```diff
@@ -1749,21 +1813,21 @@ class Player:
...
-			events += move_in_direction_by_amount(false, amount, true, "close")
+			events += move_in_direction_by_amount(false, amount, true, -1, "close")
 		else:
-			events += move_in_direction_by_amount(true, amount, true, "close")
+			events += move_in_direction_by_amount(true, amount, true, -1, "close")
 
 		return events
 
-	func advance(amount):
+	func advance(amount, stop_on_space):
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:1822]: There is only 1 place that calls this, so it might as well not be optional

File: scenes/game/local_game.gd (hunk 1 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:5851]: I think we just have to do a survey of all the things a given character that has extra attacks can boost.

File: scenes/game/local_game.gd (hunk 2 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:5962]: yes, this was me being confused about a warning, which was actually a real bug.
So, if you use events += [] inside this function, this does a new assignment and loses the reference to events and doesn't actually add to the array.
So if we ever aren't returning events from a function we pass events into, it needs to use append_array/append.
Potentially we should go change this everywhere anyway.

File: test/unit/test_randomai.gd (hunk 4 above)
Comment (daniel-k-taylor) [reply test/unit/test_randomai.gd:1]: Sure, we could have a test that just runs every character against every other character.
Probably take a while to run though, but it would have at least caught stuff like hazama vs critical

**Review by Mach131 — APPROVED** [review]
